jetty-io/src/test/java/org/eclipse/jetty/io/DefaultRetainableByteBufferPoolTest.java

```diff
@@ -0,0 +1,218 @@
+        assertThat(pool.getDirectMemory(), is(30L));
+        assertThat(pool.getAvailableDirectByteBufferCount(), is(2L));
+        assertThat(pool.getAvailableDirectMemory(), is(30L));
+    }
+
+    @Test
+    public void testFactorAndCapacity()
+    {
+        DefaultRetainableByteBufferPool pool = new DefaultRetainableByteBufferPool(10, 10, 20, Integer.MAX_VALUE);
+
+        pool.acquire(1, true);  // not pooled, < minCapacity
+        pool.acquire(10, true); // pooled
+        pool.acquire(20, true); // pooled
+        pool.acquire(30, true); // not pooled, > maxCapacity
+
+        assertThat(pool.getDirectByteBufferCount(), is(2L));
+        assertThat(pool.getDirectMemory(), is(30L));
+        assertThat(pool.getAvailableDirectByteBufferCount(), is(0L));
+        assertThat(pool.getAvailableDirectMemory(), is(0L));
+    }
+
+    @Test
+    public void testClearUnlinksLeakedBuffers()
+    {
+        DefaultRetainableByteBufferPool pool = new DefaultRetainableByteBufferPool();
+
+        pool.acquire(10, true);
+        pool.acquire(10, true);
+
+        assertThat(pool.getDirectByteBufferCount(), is(2L));
+        assertThat(pool.getDirectMemory(), is(2048L));
+        assertThat(pool.getAvailableDirectByteBufferCount(), is(0L));
+        assertThat(pool.getAvailableDirectMemory(), is(0L));
+
+        pool.clear();
+
+        assertThat(pool.getDirectByteBufferCount(), is(0L));
+        assertThat(pool.getDirectMemory(), is(0L));
+        assertThat(pool.getAvailableDirectByteBufferCount(), is(0L));
+        assertThat(pool.getAvailableDirectMemory(), is(0L));
+    }
+
+    @Test
+    public void testAcquireRelease()
+    {
+        DefaultRetainableByteBufferPool pool = new DefaultRetainableByteBufferPool();
+
+        for (int i = 0; i < 3; i++)
+        {
+            {
+                RetainableByteBuffer buffer = pool.acquire(10, true);
+                assertThat(buffer, is(notNullValue()));
+                RetainableByteBuffer buffer2 = pool.acquire(10, true);
+                assertThat(buffer2, is(notNullValue()));
+                buffer.release();
+                buffer2.release();
+            }
+            {
+                RetainableByteBuffer buffer = pool.acquire(16385, true);
+                assertThat(buffer, is(notNullValue()));
+                buffer.release();
+            }
+            {
+                RetainableByteBuffer buffer = pool.acquire(32768, true);
+                assertThat(buffer, is(notNullValue()));
+                buffer.release();
+            }
+            {
+                RetainableByteBuffer buffer = pool.acquire(32768, false);
+                assertThat(buffer, is(notNullValue()));
+                buffer.release();
+            }
+        }
+
+        assertThat(pool.getDirectByteBufferCount(), is(4L));
+        assertThat(pool.getHeapByteBufferCount(), is(1L));
+        assertThat(pool.getDirectMemory(), is(52224L));
+        assertThat(pool.getHeapMemory(), is(32768L));
```

Comment:
Too many magic values. You can allocate variables and perform a sum here, e.g. `capacity1=1024` but then `acquire(capacity1/10, ...)`, etc.

Reply:
Agreed. I cleaned up this test to get rid of magic constants. I also added a few more assertions that make it more obvious what buffer capacity is expected from a certain allocation.